iris-mpc-cpu/src/execution/hawk_main/intra_batch.rs

```diff
@@ -0,0 +1,126 @@
+    ];
+
+    let per_session = |batch: Batch| {
+        let search_queries = search_queries[batch.i_eye].clone();
+        let session = sessions[batch.i_eye][batch.i_session].clone();
+        async move {
+            let mut session = session.write().await;
+            per_session(&search_queries, &mut session, batch).await
+        }
+    };
+
+    let results = batches
+        .into_iter()
+        .map(per_session)
+        .map(tokio::spawn)
+        .collect::<JoinAll<_>>()
+        .await;
+
+    aggregate_results(n_requests, results)
+}
+
+async fn per_session(
+    search_queries: &VecRequests<VecRots<QueryRef>>,
+    session: &mut HawkSession,
+    batch: Batch,
+) -> Vec<IsMatch> {
+    // Enumerate the pairs of requests.
+    // These are unordered pairs: if we do (i, j) we skip (j, i).
+    let pairs = batch
+        .tasks
+        .into_iter()
+        .flat_map(|task| {
+            (0..task.i_request).map(move |other_request| IsMatch {
+                eye: batch.i_eye,
+                task,
+                other_request,
+            })
+        })
+        .collect_vec();
+
+    // Compare the rotated and processed irises of one request, to the centered unprocessed iris of the other request.
+    let query_pairs = pairs
+        .iter()
+        .map(|pair| {
+            (
+                &search_queries[pair.task.i_request][pair.task.i_rotation].processed_query,
+                &search_queries[pair.other_request].center().query,
+            )
+        })
+        .collect_vec();
+
+    let distances = session
+        .aby3_store
+        .eval_pairwise_distances(query_pairs)
+        .await;
+    let distances = session.aby3_store.lift_distances(distances).await.unwrap();
+    let is_matches = session.aby3_store.is_match_batch(&distances).await;
+
+    izip!(pairs, is_matches)
+        .filter_map(|(pair, is_match)| is_match.then_some(pair))
+        .collect_vec()
+}
+
+struct IsMatch {
+    eye: usize,
+    task: Task,
+    other_request: usize,
+}
+
+fn aggregate_results(
+    n_requests: usize,
+    results: Vec<Result<Vec<IsMatch>, JoinError>>,
+) -> Result<VecRequests<Vec<usize>>> {
+    let mut join = HashMap::new();
+
+    // For each pair of request, reduce the result of all rotations with boolean ANY.
+    for batch in results {
+        for match_result in batch? {
+            let request_pair = (match_result.task.i_request, match_result.other_request);
+            let eyes_match = join.entry(request_pair).or_insert([false, false]);
+            eyes_match[match_result.eye] = true;
+        }
+    }
+
+    let mut match_lists = vec![Vec::new(); n_requests];
+
+    // Find pairs with left AND right match.
+    for ((i_request, j_request), [left, right]) in join {
+        if left && right {
+            // The pair of matching requests track each other's index.
+            match_lists[i_request].push(j_request);
+            match_lists[j_request].push(i_request);
```

Comment:
If i recall correctly, this is not the current behavior in the batch matching in GPU. I think atm if we have x in the batch twice, [x, a,b,c,d,e,f,x], then we report a batch match for the second x but not for the first one, since it is fine for the first one to be unique in that case. This is the case, right @philsippl?

Reply:
That makes sense. See the latest changes around `earlier_request` and around `HawkResult::new`.

Reply:
What do we do if there is a chain of matches though? Say request `J` is a duplicate of request `I`, but `I` will also not be inserted (for any reason), then should `J` be inserted 🙃 ?

Reply:
I guess atm this is handled by the assumption that if J is not inserted due to a reason then I will also not be inserted due to the same reason. (I guess in theory one could craft iris codes such that J matches DB, I does not, but J matches I. This is probably an edge case and the default behavior of not inserting I in this case is also fine?)

Reply:
Yes that sounds sensible.

Reply:
yes, it's as @dkales describes.